Documentation/development-flow.md

```diff
@@ -0,0 +1,115 @@
+# Prerequisites
+1. [GO 1.8](https://golang.org/dl/) or greater installed
+2. Git client installed
+3. Github account
+
+# Initial Setup
+
+## Create a Fork
+
+From your browser navigate to [http://github.com/rook/rook](http://github.com/rook/rook) and click the "Fork" button. 
+
+## Clone Your Fork
+
+Open a console window and do the following;
+```bash
+# Create the rook repo path
+mkdir -p $GOPATH/src/github.com/rook
+
+# Navigate to the local repo path and clone your fork
+cd $GOPATH/src/github.com/rook
+
+# Clone your fork, where <user> is your github account name
+git clone git@github.com:<user>/rook.git
+
+cd rook
+```
+
+## Add Upstream Remote
+First you will need to add the upstream remote to your local git:
+```bash
+# Add 'upstream' to the list of remotes
+git remote add upstream https://github.com/rook/rook.git
+
+# Verify the remote was added
+git remote -v
+```
+Now you should have at least `origin` and `upstream` remotes. You can also add other remotes to collaborate with other contributors.
+
+
+# Development
+To add a feature or to make a bug fix, you will need to create a branch in your fork and then submit a pull request (PR) from the branch.
+
+## Create a Branch
+From a console, create a new branch based on your fork and start working on it:
+```bash
+# Checkout the master branch - you want your new branch to come from master
+git checkout master
+
+# Create a new branch with a simple, but descriptive name. Generally it will be two to three words separated by dashes and without numbers.
+git checkout -b feature-name
+```
+
+Now you are ready to make the changes and commit to your branch.
+
+## Updating Your Fork
+During the development lifecycle, you will need to keep up-to-date with the latest upstream master. As others on the team push changes, you will need to `rebase` your commits on top of the latest. This avoids unnecessary merge commits and keeps the commit history clean. 
+
+Whenever you need to update your local repository, you never want to merge. You **always** will rebase. Otherwise you will end up with merge commits in the git history. If you have any modified files, you will first have to stash them (`git stash save "<some description>"`).
+
+```bash
+git rebase upstream/master
+```
+
+Rebasing is a very powerful feature of Git. You need to understand how it works or else you will risk losing your work. Read about it in the [Git documentation](https://git-scm.com/docs/git-rebase), it will be well worth it. In a nutshell, rebasing does the following:
+- "Unwinds" your local commits. Your local commits are removed temporarily from the history.
+- The latest changes from upstream are added to the history
+- Your local commits are re-applied one by one
+- If there are merge conflicts, you will be prompted to fix them before continuing. Read the output closely. It will tell you how to complete the rebase.
+- When done rebasing, you will see all of your commits in the history.
```

Comment:
Add a reminder that if you used `git stash` to save uncommitted changes, you should now do a `git stash pop` to restore them.